### PR TITLE
[Sparse] Link to DGL

### LINKED ISSUE #4852: [Sparse] Link spares C++ library to DGL
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description

Resolve #4852.

1. Build C++ sparse library in DGL CMakeLists.txt and link it to DGL. 
2. Move the definitions of sparse formats to a separate file.
3. Add a CSRToCOO conversion using DGL implementation with test.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by czkkkkkk** [comment]
When linking dgl_sparse to both DGL and Libtorch, I met two major problems.
1. Diamond dependency for tensorpipe and logger. Both DGL and Pytorch have their own source code for tensorpipe and logger, leading to the same definition under the same namespace. Tensorpipe is still fine because it only raise some redefinition warnings during compilation. But the conflict of `dmlc/logging.h` and `torch/include/c10/util/logging_is_not_google_glog.h` could make the compilation fail. An example is when DGL uses the `CHECK_NOTNULL` from dmlc/logging.h, it will point to the `CHECK_NOTNULL` in Pytorch, which has a different signature.
image
2. Ambiguity of kInt, kFloat, etc. Both DGL (`dgl/runtime/c_runtime_api.h`) and Pytorch have these enumerations. We may need change our enum to `enum class {kInt, kFloat ...}`.

Changing the include order could magically solve these two issues. But we need to address them permanently. Do you have any suggestion? @jermainewang @BarclayII

File: dgl_sparse/include/sparse/sparse_format.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/**
+ *  Copyright (c) 2022 by Contributors
+ * @file sparse/sparse_format.h
+ * @brief DGL C++ sparse format header
```
Comment (frozenbugs) [inline dgl_sparse/include/sparse/sparse_format.h:4]: nit: always end with dot.

For all the comments below as well.

File: dgl_sparse/src/utils.h [hunk 2]
```diff
@@ -45,6 +48,16 @@ inline static void ElementwiseOpSanityCheck(
       << ", " << B->shape()[1] << "])";
 }
 
+/** @brief Convert a DGL array to Torch tensor. */
```
Comment (frozenbugs) [inline dgl_sparse/src/utils.h:51]: /** @brief Convert a Torch tensor to DGL array. */

File: dgl_sparse/src/utils.h [hunk 3]
```diff
@@ -45,6 +48,16 @@ inline static void ElementwiseOpSanityCheck(
...
+inline static runtime::NDArray TorchTensorToDGLArray(torch::Tensor tensor) {
+  return runtime::DLPackConvert::FromDLPack(at::toDLPack(tensor));
+}
+
+/** @brief Convert a Torch tensor to DGL array. */
```
Comment (frozenbugs) [inline dgl_sparse/src/utils.h:56]: ditto.

File: dgl_sparse/src/sparse_format.cc [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    int64_t num_rows, int64_t num_cols, const std::shared_ptr<CSR> csr) {
+  auto indptr = TorchTensorToDGLArray(csr->indptr);
+  auto indices = TorchTensorToDGLArray(csr->indices);
+  bool data_as_order = false;
+  runtime::NDArray data = aten::NullArray();
+  if (csr->value_indices.has_value()) {
+    data_as_order = true;
+    data = TorchTensorToDGLArray(csr->value_indices.value());
+  }
+  auto dgl_csr = aten::CSRMatrix(num_rows, num_cols, indptr, indices, data);
```
Comment (jermainewang) [inline dgl_sparse/src/sparse_format.cc:27]: * Perhaps we could add some util functions like `CSRToOldDGLCSR`, `OldDGLCSRToCSR` as they will be quite commonly used in other operators.
* I also suggest to add the `sorted` related flags to the new CSR/COO structure. They are performance critical for certain operations.

File: tests/pytorch/mock_sparse2/test_sparse_matrix.py [hunk 5]
```diff
@@ -89,3 +89,30 @@ def test_create_from_csc(dense_dim, indptr, indices, shape):
     assert torch.allclose(mat_indptr, indptr)
     assert torch.allclose(mat_indices, indices)
     assert torch.allclose(mat_val, val)
+
+@pytest.mark.parametrize("dense_dim", [None, 4])
+@pytest.mark.parametrize("indptr", [[0, 0, 1, 4], (0, 1, 2, 4)])
+@pytest.mark.parametrize("indices", [(0, 1, 2, 3), (1, 4, 3, 2)])
+@pytest.mark.parametrize("shape", [None, (3, 5)])
+def test_csr_to_coo(dense_dim, indptr, indices, shape):
```
Comment (jermainewang) [inline tests/pytorch/mock_sparse2/test_sparse_matrix.py:97]: Could you add tests on GPU too? For now, you could use the `ctx()` object from the test backend to get the current device being tested.

**Review by jermainewang — APPROVED** [review]
Mostly good. Some minor comments.

File: tests/pytorch/mock_sparse2/test_sparse_matrix.py (hunk 5 above)
Comment (czkkkkkk) [reply tests/pytorch/mock_sparse2/test_sparse_matrix.py:99]: Added

File: dgl_sparse/src/sparse_format.cc (hunk 4 above)
Comment (czkkkkkk) [reply dgl_sparse/src/sparse_format.cc:27]: OK. I will have these two things in the next PR. This PR is mainly for linking to DGL.

**Review by BarclayII — APPROVED** [review]
I like this solution.

**Comment by yaox12** [comment]
Better to add options for `BUILD_SPARSE` and `TORCH_PYTHON_INTERPS` in
https://github.com/dmlc/dgl/blob/3103f8cb362033d9b1543f4bd4b2bd9ac87a61d4/CMakeLists.txt#L38-L39

**Comment by czkkkkkk** [comment]
> Better to add options for `BUILD_SPARSE` and `TORCH_PYTHON_INTERPS` in
> 
> https://github.com/dmlc/dgl/blob/3103f8cb362033d9b1543f4bd4b2bd9ac87a61d4/CMakeLists.txt#L38-L39
You are right. Will make a PR to include them.

**Pull request merged**